File: src/main/scala/com/minalien/mffs/tiles/TileEntityExtractor.scala
package com.minalien.mffs.tiles

import cofh.api.energy.IEnergyHandler
import com.minalien.mffs.api.{EnumUpgradeType, IPowerLinkItem}
import com.minalien.mffs.containers.ContainerForceEnergyExtractor
import com.minalien.mffs.core.{LinkGrid, MFFSConfig}
import com.minalien.mffs.items.{ItemUpgrade, ItemForciumCell, ItemForcium}
import com.minalien.mffs.utils.NBTUtils
import net.minecraft.entity.player.InventoryPlayer
import net.minecraft.inventory.Container
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.server.gui.IUpdatePlayerListBox
import net.minecraft.util.{ChatComponentText, EnumFacing, IChatComponent}
import net.minecraftforge.common.util.Constants

/**
 * Created by Katrina on 15/12/2014.
 */
class TileEntityExtractor extends TileEntityFEPoweredMachine with IUpdatePlayerListBox with IEnergyHandler{
  val inventory: Array[ItemStack] = new Array[ItemStack](5);
  var workmode: Int = 0
  var WorkEnergy: Int = 0
  var MaxWorkEnergy: Int = 4000
  var ForceEnergybuffer: Int = 0
  var MaxForceEnergyBuffer: Int = 1000000
  //                              100000
  var WorkCycle: Int = 0
  var workTicker: Int = 20
  var workdone: Int = 0
  var maxworkcycle: Int = 125
  var capacity: Int = 0


  def getCapacity():Int=capacity

  def setCapacity(cap:Int)=
  {
    if(this.capacity!=cap)
    {
      this.capacity=cap
      worldObj.markBlockForUpdate(pos)
    }
  }
  def getWorkDone()=workdone

  def setWorkDone(workdone:Int)=
  {
    if(this.workdone!=workdone)
    {
      this.workdone=workdone
      worldObj.markBlockForUpdate(pos)
    }
  }

  def getWorkCycle()=WorkCycle

  def setWorkCycle(WorkCycle:Int)=
  {
    if(this.WorkCycle!=WorkCycle)
    {
      this.WorkCycle=WorkCycle
      worldObj.markBlockForUpdate(pos)
    }
  }
  def getForceEnergyBuffer() = ForceEnergybuffer
  def setForceEnergyBuffer(force:Int) =
  {
    if(this.ForceEnergybuffer!=force)
    {
      this.ForceEnergybuffer=force
      worldObj.markBlockForUpdate(pos)
    }
  }

  override def dropPlugins: Unit =
  {
    for(i <- 0 until inventory.length)
      dropPlugins(i,this)
  }
  def checkSlots(init:Boolean)=
  {
    if(getStackInSlot(2) ne null)
    {
      getStackInSlot(2).getItem match
      {
         case s:ItemUpgrade=>
          s.getTypeUpgrade() match {
            case EnumUpgradeType.CapacitorUpgradeCapacity =>
              MaxForceEnergyBuffer= 1000000 +(getStackInSlot(2).stackSize * 1000000)
            case _ => MaxForceEnergyBuffer = 1000000
          }
         case _ => MaxForceEnergyBuffer= 1000000
      }
    }
    else
      MaxForceEnergyBuffer= 1000000

    if(getStackInSlot(3) ne null)
    {
      getStackInSlot(3).getItem match {
        case s: ItemUpgrade =>
          s.getTypeUpgrade() match {
            case EnumUpgradeType.ExtractorBooster => workTicker = 20 - getStackInSlot(3).stackSize
            case _ => workTicker = 20
          }
        case _ => workTicker=20
      }
    }
    else
      workTicker=20

    if(getStackInSlot(4) ne null)
    {
      getStackInSlot(4).getItem match
      {
        case ItemForciumCell => {
          workmode=1
          MaxWorkEnergy=200000

        }
      }
    }
    else
    {
      workmode=0
      MaxWorkEnergy=4000
    }
  }

  def hasPowerToConvert:Boolean=
  {
    if(WorkEnergy>=MaxWorkEnergy-1)
    {
      WorkEnergy=0
      true
    }
    else
      false
  }

  def hasFreeForceEnergyStorage:Boolean=
  {
    if(this.MaxForceEnergyBuffer> this.ForceEnergybuffer)
      true
    else
      false
  }
  def hasStuffToConvert:Boolean= {
    if (WorkCycle > 0)
      true
    else if(MFFSConfig.General.adventureMode)
    {
      maxworkcycle=MFFSConfig.Machines.ForciciumCellWorkCycle
      setWorkCycle(maxworkcycle)
      true

    } else if(getStackInSlot(0)!=null)
    {
      getStackInSlot(0).getItem match
      {
        case ItemForcium =>
        {
          maxworkcycle=MFFSConfig.Machines.ForciciumWorkCycle
          setWorkCycle(maxworkcycle)
          decrStackSize(0,1)
          true
        }
        case ItemForciumCell =>
        {
          if(ItemForciumCell.useForceium(1,getStackInSlot(0)))
          {
            maxworkcycle=MFFSConfig.Machines.ForciciumCellWorkCycle
            setWorkCycle(maxworkcycle)
            true
          }
          else
            false
        }
        case _ => false
      }
    }
    else
      false
  }
  def transferForceEnergy() =
  {
    if(this.ForceEnergybuffer>0)
    {
      if(this.hasPowerSource)
      {
        val PowerTransferRate=this.getMaximumPower/120
        val freeAmount=this.getMaximumPower-this.getAvailablePower
        if(this.ForceEnergybuffer>freeAmount)
        {
          if(freeAmount>PowerTransferRate)
          {
            emitPower(PowerTransferRate,false)
            this.setForceEnergyBuffer(ForceEnergybuffer-PowerTransferRate)
          }
          else
          {
            emitPower(freeAmount,false)
            this.setForceEnergyBuffer(ForceEnergybuffer-freeAmount)
          }
        }
        else
        {
          if(freeAmount>ForceEnergybuffer)
          {
            emitPower(ForceEnergybuffer,false)
            this.setForceEnergyBuffer(0)
          }
          else
          {
            emitPower(freeAmount,false)
            this.setForceEnergyBuffer(ForceEnergybuffer-freeAmount)
          }
        }
      }
    }
  }

  override def getmaxSwitchMode=3
  override def getminSwitchMode=1

  override def update(): Unit =
  {
    if(worldObj.isRemote==false)
    {
      if(this.init)
        checkSlots(true)

      if(getSwitchMode==1)
        if(!getSwitchValue && isRedstoneSignal)
          toggleSwitchValue
      if(getSwitchMode==1)
        if(getSwitchValue && !isRedstoneSignal)
          toggleSwitchValue
      if(!isActive && getSwitchValue)
        this.activate()
      if(isActive && !getSwitchValue)
        this.deactivate()
      // TODO: REmove this once we have some power API
      WorkEnergy=MaxWorkEnergy
      if(this.getTicker >=workTicker)
      {
        checkSlots(false)
        if(workmode==0 && isActive)
        {
          if(this.workdone!=WorkEnergy*100/MaxWorkEnergy)
            setWorkDone(WorkEnergy*100/MaxWorkEnergy)

          if(workdone>100)
            setWorkDone(workdone=100)

          if(getCapacity()!=(ForceEnergybuffer*100)/MaxForceEnergyBuffer)
          {
            setCapacity((ForceEnergybuffer*100)/MaxForceEnergyBuffer)
          }
          if(this.hasFreeForceEnergyStorage && this.hasStuffToConvert)
          {
            if(this.hasPowerToConvert)
            {
              setWorkCycle(WorkCycle-1)
              this.setForceEnergyBuffer(ForceEnergybuffer+MFFSConfig.Machines.ExtractorPassForceEnergyGenerate)
            }
          }
          transferForceEnergy()
          this.setTicker(0)
        }
        if(workmode==1 && isActive)
        {
          if(this.workdone!=WorkEnergy*100/MaxWorkEnergy)
            setWorkDone(WorkEnergy*100/MaxWorkEnergy)

          if(workdone>100)
            setWorkDone(workdone=100)
          if(ItemForciumCell.getForceciumlevel(getStackInSlot(4))<ItemForciumCell.getMaxForceciumlevel)
          {
            if(this.hasPowerToConvert && isActive)
            {
              ItemForciumCell.setForceciumlevel(getStackInSlot(4),ItemForciumCell.getForceciumlevel(getStackInSlot(4))+1)
            }
          }
          this.setTicker(0)
        }
      }
      this.setTicker(this.getTicker+1)
    }
    super.update()
  }

  override def getContainer(inventoryplayer: InventoryPlayer): Container = new ContainerForceEnergyExtractor(inventoryplayer.player,this)


  override def getStackInSlot(index: Int): ItemStack = inventory(index)


  override def getName: String = "Extractor"


  override def hasCustomName: Boolean = false

  override def getDisplayName: IChatComponent = new ChatComponentText("Extractor")


  override def getSizeInventory: Int = inventory.length


  override def setInventorySlotContents(index: Int, stack: ItemStack): Unit =
  {
    this.inventory(index) = stack

    if (stack != null && stack.stackSize > this.getInventoryStackLimit) {
      stack.stackSize = this.getInventoryStackLimit
    }

    this.markDirty
  }


  override def decrStackSize(index: Int, count: Int): ItemStack =
  {
    if (this.inventory(index) != null) {
      var itemstack: ItemStack = null
      if (this.inventory(index).stackSize <= count) {
        itemstack = this.inventory(index)
        this.inventory(index) = null
        this.markDirty
        return itemstack
      }
      else {
        itemstack = this.inventory(index).splitStack(count)
        if (this.inventory(index).stackSize == 0) {
          this.inventory(index) = null
        }
        this.markDirty
        return itemstack
      }
    }
    else {
      return null
    }
  }


  override def isItemValid(par1ItemStack: ItemStack, Slot: Int): Boolean =
  {
    Slot match
    {
      case 0=>
      {
        par1ItemStack.getItem match
        {
          case ItemForciumCell => getStackInSlot(4)==null
          case ItemForcium => getStackInSlot(4)==null
          case _ => false
        }
      }
      case 1 =>
      {
        par1ItemStack.getItem match {
          case item: IPowerLinkItem => true
          case _ => false
        }
      }
      case 2 =>
      {
        par1ItemStack.getItem match
        {
          case item: ItemUpgrade => item.getTypeUpgrade() match { case EnumUpgradeType.CapacitorUpgradeCapacity => true case _ => false}
          case _ => false
        }
      }
      case 3 =>
        par1ItemStack.getItem match
        {
          case item: ItemUpgrade => item.getTypeUpgrade() match { case EnumUpgradeType.ExtractorBooster=> true case _ => false}
          case _ => false
        }
      case 4 =>
      {
        par1ItemStack.getItem match
        {
          case ItemForciumCell => true
          case _ => false
        }
      }
    }
  }


  override def getSlotStackLimit(slt: Int): Int = {
    slt match
    {
      case 0 =>64
      case 1 => 1
      case 2 => 9
      case 3 => 19
      case 4 => 1
      case _ => 1
    }
  }


  override def receiveEnergy(from: EnumFacing, maxReceive: Int, simulate: Boolean): Int =
  {
    val freeSpace:Double=MaxWorkEnergy-WorkEnergy
    if(freeSpace>= maxReceive)
    {
      if(!simulate)
        WorkEnergy=WorkEnergy+maxReceive
      0
    }
    else
    {
      WorkEnergy=MaxWorkEnergy
      (maxReceive-freeSpace).toInt
    }
  }

  override def extractEnergy(from: EnumFacing, maxExtract: Int, simulate: Boolean): Int = 0

  override def getEnergyStored(from: EnumFacing): Int = WorkEnergy

  override def getMaxEnergyStored(from: EnumFacing): Int = MaxWorkEnergy

  override def canConnectEnergy(from: EnumFacing): Boolean = true



  override def getPowerLinkStack: ItemStack = this.getStackInSlot(getPowerlinkSlot)

  override def getPowerlinkSlot: Int = 1


  override def getLinkedSecurityStation: MFFSMachine = null // TODO: Finish once have security station



  override def isItemValidForSlot(index: Int, stack: ItemStack): Boolean = isItemValid(stack,index)



  override def getSlotsForFace(side: EnumFacing): Array[Int] =
  {
    side match
    {
      case EnumFacing.UP=> Array(0)
      case EnumFacing.DOWN=> Array(1)
      case EnumFacing.NORTH=>Array(4)
      case EnumFacing.EAST=>Array(4)
      case EnumFacing.SOUTH=>Array(4)
      case EnumFacing.WEST=>Array(4)
      case _ => Array()
    }
  }

  override def canExtractItem(index: Int, stack: ItemStack, direction: EnumFacing): Boolean =true

  override def canInsertItem(index: Int, itemStackIn: ItemStack, direction: EnumFacing): Boolean = true

  override def clear(): Unit = {

    for (i <- 0 until this.inventory.length) {
      this.inventory(i) = null
    }


  }



  override def readExtraNBT(nbttagcompound: NBTTagCompound): Unit =
  {
    super.readExtraNBT(nbttagcompound)
    ForceEnergybuffer=nbttagcompound.getInteger("ForceEnergyBuffer")
    WorkEnergy=nbttagcompound.getInteger("WorkEnergy")
    WorkCycle=nbttagcompound.getInteger("WorkCycle")
    NBTUtils.inventoryFromNBT(inventory,nbttagcompound.getTagList("Items",Constants.NBT.TAG_COMPOUND))
  }
  override def readTransientNBT(nbttagcompound:NBTTagCompound)=
  {
    capacity=nbttagcompound.getInteger("Capacity")
    workdone=nbttagcompound.getInteger("WorkDone")
  }
  override def writeExtraNBT(nbttagcompound: NBTTagCompound): Unit =
  {
    super.writeExtraNBT(nbttagcompound)
    nbttagcompound.setInteger("ForceEnergyBuffer",ForceEnergybuffer)
    nbttagcompound.setInteger("WorkEnergy",WorkEnergy)
    nbttagcompound.setInteger("WorkCycle",WorkCycle)
    nbttagcompound.setTag("Items",NBTUtils.NBTFromInventory(inventory))
  }

  override def writeTransientNBT(nbttagcompound:NBTTagCompound)=
  {
    nbttagcompound.setInteger("Capacity",capacity)
    nbttagcompound.setInteger("WorkDone",workdone)
  }

  override def invalidate(): Unit =
  {
    LinkGrid.getWorldNet(worldObj).Extractors.remove(DeviceID)
    super.invalidate()
  }
}

File: src/main/java/com/nekokittygames/mffs/common/tileentity/TileEntityExtractor.java
/*  
    Copyright (C) 2012 Thunderdark

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
    
    Contributors:
    
    Matchlighter
    Thunderdark 

 */

package com.nekokittygames.mffs.common.tileentity;



import cofh.redstoneflux.api.IEnergyReceiver;
import com.nekokittygames.mffs.api.IPowerLinkItem;
import com.nekokittygames.mffs.common.Linkgrid;
import com.nekokittygames.mffs.common.ModularForceFieldSystem;
import com.nekokittygames.mffs.common.compat.TeslaCap;
import com.nekokittygames.mffs.common.container.ContainerForceEnergyExtractor;
import com.nekokittygames.mffs.common.item.*;
import ic2.api.energy.event.EnergyTileLoadEvent;
import ic2.api.energy.event.EnergyTileUnloadEvent;
import ic2.api.energy.tile.IEnergyEmitter;
import ic2.api.energy.tile.IEnergySink;
import net.darkhax.tesla.capability.TeslaCapabilities;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.inventory.Container;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.fml.common.Loader;
import net.minecraftforge.fml.common.Optional;

import java.util.LinkedList;
import java.util.List;

@Optional.InterfaceList({@Optional.Interface(modid = "ic2",iface = "ic2.api.energy.tile.IEnergySink"),
							@Optional.Interface(modid = "ic2",iface = "ic2.api.energy.tile.IEnergyEmitter")})
public class TileEntityExtractor extends TileEntityFEPoweredMachine implements
		IEnergySink, IEnergyReceiver {
	private int workmode = 0;

	protected int WorkEnergy;
	protected int MaxWorkEnergy;
	private int ForceEnergybuffer;
	private int MaxForceEnergyBuffer;
	private int WorkCylce;
	private int workTicker;
	private int workdone;
	private int maxworkcylce;
	private int capacity;
	//private IPowerEmitter powerEmitter;
	private boolean addedToEnergyNet;

	private TeslaCap cap;

	public TileEntityExtractor() {
		super(5);
		WorkEnergy = 0;
		MaxWorkEnergy = 4000;
		ForceEnergybuffer = 0;
		MaxForceEnergyBuffer = 1000000;
		WorkCylce = 0;
		workTicker = 20;
		maxworkcylce = 125;
		capacity = 0;
		addedToEnergyNet = false;
		if(Loader.isModLoaded("tesla"))
			cap=new TeslaCap(this);


	}



	public int getCapacity() {
		return capacity;
	}

	public void setCapacity(int Capacity) {
		if (this.capacity != Capacity) {
			this.capacity = Capacity;
		}
	}

	public int getMaxworkcylce() {
		return maxworkcylce;
	}

	public void setMaxworkcylce(int maxworkcylce) {
		this.maxworkcylce = maxworkcylce;
	}

	public int getWorkdone() {
		return workdone;
	}

	public void setWorkdone(int workdone) {
		if (this.workdone != workdone) {
			this.workdone = workdone;

		}
	}

	public int getWorkTicker() {
		return workTicker;
	}

	public void setWorkTicker(int workTicker) {
		this.workTicker = workTicker;
	}

	public int getMaxForceEnergyBuffer() {
		return MaxForceEnergyBuffer;
	}

	public void setMaxForceEnergyBuffer(int maxForceEnergyBuffer) {
		MaxForceEnergyBuffer = maxForceEnergyBuffer;
		this.markDirty();
		getWorld().notifyBlockUpdate(pos, getWorld().getBlockState(pos), getWorld().getBlockState(pos), 2);
		getWorld().markBlockRangeForRenderUpdate(pos,pos);
		//todo: make more packet friendly
	}

	public int getForceEnergybuffer() {
		return ForceEnergybuffer;
	}

	public void setForceEnergybuffer(int forceEnergybuffer) {
		ForceEnergybuffer = forceEnergybuffer;
		this.markDirty();
		getWorld().notifyBlockUpdate(pos, getWorld().getBlockState(pos), getWorld().getBlockState(pos), 2);
		getWorld().markBlockRangeForRenderUpdate(pos,pos);
	}

	public void setWorkCylce(int i) {
		if (this.WorkCylce != i) {
			this.WorkCylce = i;
		}
		this.markDirty();
		getWorld().notifyBlockUpdate(pos, getWorld().getBlockState(pos), getWorld().getBlockState(pos), 2);
		getWorld().markBlockRangeForRenderUpdate(pos,pos);
	}

	public int getWorkCylce() {
		return WorkCylce;
	}

	public int getWorkEnergy() {
		return WorkEnergy;
	}

	public void setWorkEnergy(int workEnergy) {
		WorkEnergy = workEnergy;
		this.markDirty();
		getWorld().notifyBlockUpdate(pos, getWorld().getBlockState(pos), getWorld().getBlockState(pos), 2);
		getWorld().markBlockRangeForRenderUpdate(pos,pos);
	}

	public int getMaxWorkEnergy() {
		return MaxWorkEnergy;
	}

	public void setMaxWorkEnergy(int maxWorkEnergy) {
		MaxWorkEnergy = maxWorkEnergy;
		this.markDirty();
		getWorld().notifyBlockUpdate(pos, getWorld().getBlockState(pos), getWorld().getBlockState(pos), 2);
		getWorld().markBlockRangeForRenderUpdate(pos,pos);
	}

	@Override
	public void dropPlugins() {
		for (int a = 0; a < this.inventory.size(); a++) {
			dropPlugins(a);
		}
	}



	public void checkslots(boolean init) {

		if (!getStackInSlot(2).isEmpty()) {
			if (getStackInSlot(2).getItem() == ModItems.UPGRADE_CAPACITY) {
				setMaxForceEnergyBuffer(1000000 + (getStackInSlot(2).getCount() * 100000));
			} else {
				setMaxForceEnergyBuffer(1000000);
			}
		} else {
			setMaxForceEnergyBuffer(1000000);
		}

		if (!getStackInSlot(3).isEmpty()) {
			if (getStackInSlot(3).getItem() == ModItems.EXTRACTOR_UPGRADE_BOOSTER) {
				setWorkTicker(20 - getStackInSlot(3).getCount());
			} else {
				setWorkTicker(20);
			}
		} else {
			setWorkTicker(20);
		}

		if (!getStackInSlot(4).isEmpty()) {
			if (getStackInSlot(4).getItem() == ModItems.FORCICIUM_CELL) {
				workmode = 1;
				setMaxWorkEnergy(200000);
			}
		} else {
			workmode = 0;
			setMaxWorkEnergy(4000);
		}
	}

	private boolean hasPowertoConvert() {
		if (WorkEnergy >= MaxWorkEnergy - 1) {
			setWorkEnergy(0);
			return true;
		}
		return false;
	}

	private boolean hasfreeForceEnergyStorage() {
		if (this.MaxForceEnergyBuffer > this.ForceEnergybuffer)
			return true;
		return false;
	}

	private boolean hasStufftoConvert() {
		if (WorkCylce > 0) {
			return true;
		} else {
			if (ModularForceFieldSystem.adventureMapMode) {
				setMaxworkcylce(ModularForceFieldSystem.ForciciumCellWorkCycle);
				setWorkCylce(getMaxworkcylce());
				return true;
			}

			if (!getStackInSlot(0).isEmpty()) {
				if (getStackInSlot(0).getItem() == ModItems.FORCICIUM) {
					setMaxworkcylce(ModularForceFieldSystem.ForciciumWorkCycle);
					setWorkCylce(getMaxworkcylce());
					decrStackSize(0, 1);
					return true;
				}

				if (getStackInSlot(0).getItem() == ModItems.FORCICIUM_CELL) {
					if (((ItemForcicumCell) getStackInSlot(0).getItem())
							.useForcecium(1, getStackInSlot(0))) {
						setMaxworkcylce(ModularForceFieldSystem.ForciciumCellWorkCycle);
						setWorkCylce(getMaxworkcylce());
						return true;
					}
				}
			}
		}

		return false;
	}

	public void transferForceEnergy() {
		if (this.getForceEnergybuffer() > 0) {
			if (this.hasPowerSource()) {
				int PowerTransferrate = this.getMaximumPower() / 120;
				int freeAmount = this.getMaximumPower()
						- this.getAvailablePower();

				if (this.getForceEnergybuffer() > freeAmount) {
					if (freeAmount > PowerTransferrate) {
						emitPower(PowerTransferrate, false);
						this.setForceEnergybuffer(this.getForceEnergybuffer()
								- PowerTransferrate);

					} else {
						emitPower(freeAmount, false);
						this.setForceEnergybuffer(this.getForceEnergybuffer()
								- freeAmount);
					}
				} else {
					if (freeAmount > this.getForceEnergybuffer()) {
						emitPower(getForceEnergybuffer(), false);
						this.setForceEnergybuffer(this.getForceEnergybuffer()
								- getForceEnergybuffer());
					} else {
						emitPower(freeAmount, false);
						this.setForceEnergybuffer(this.getForceEnergybuffer()
								- freeAmount);
					}
				}
			}
		}

	}

	@Override
	public short getmaxSwitchModi() {
		return 3;
	}

	@Override
	public short getminSwitchModi() {
		return 1;
	}

	@Override
	public void update() {
		if (getWorld().isRemote == false) {

			if (init) {
				checkslots(true);
				if(addedToEnergyNet==false)
				{
					if (ModularForceFieldSystem.ic2Found)
						AddToIC2EnergyNet();
					else
						addedToEnergyNet=true;
				}

			}



			if (getSwitchModi() == 1)
				if (!getSwitchValue() && isRedstoneSignal())
					toggelSwitchValue();

			if (getSwitchModi() == 1)
				if (getSwitchValue() && !isRedstoneSignal())
					toggelSwitchValue();

			if (!isActive() && getSwitchValue())
				setActive(true);

			if (isActive() && !getSwitchValue())
				setActive(false);

			if (isActive()) {
				if (ModularForceFieldSystem.buildcraftFound)
					converMJtoWorkEnergy();
			}

			if (this.getTicker() >= getWorkTicker()) {
				checkslots(false);

				if (workmode == 0 && isActive()) {

					if (this.getWorkdone() != getWorkEnergy() * 100
							/ getMaxWorkEnergy())
						setWorkdone(getWorkEnergy() * 100 / getMaxWorkEnergy());

					if (getWorkdone() > 100) {
						setWorkdone(100);
					}

					if (this.getCapacity() != (getForceEnergybuffer() * 100)
							/ getMaxForceEnergyBuffer())
						setCapacity((getForceEnergybuffer() * 100)
								/ getMaxForceEnergyBuffer());

					if (this.hasfreeForceEnergyStorage()
							&& this.hasStufftoConvert()) {

						if (this.hasPowertoConvert()) {
							setWorkCylce(getWorkCylce() - 1);
							setForceEnergybuffer(getForceEnergybuffer()
									+ ModularForceFieldSystem.ExtractorPassForceEnergyGenerate);
						}
					}

					transferForceEnergy();

					this.setTicker((short) 0);
				}

				if (workmode == 1 && isActive()) {
					if (this.getWorkdone() != getWorkEnergy() * 100
							/ getMaxWorkEnergy())
						setWorkdone(getWorkEnergy() * 100 / getMaxWorkEnergy());

					if (((ItemForcicumCell) getStackInSlot(4).getItem())
							.getForceciumlevel(getStackInSlot(4)) < ((ItemForcicumCell) getStackInSlot(
							4).getItem()).getMaxForceciumlevel()) {

						if (this.hasPowertoConvert() && isActive()) {
							((ItemForcicumCell) getStackInSlot(4).getItem())
									.setForceciumlevel(
											getStackInSlot(4),
											((ItemForcicumCell) getStackInSlot(
													4).getItem())
													.getForceciumlevel(getStackInSlot(4)) + 1);
						}
					}

					this.setTicker((short) 0);
				}
			}

			this.setTicker((short) (this.getTicker() + 1));
		}
		super.update();
	}

	@Optional.Method(modid = "ic2")
	private void AddToIC2EnergyNet() {
		if (!getWorld().isRemote) {
			EnergyTileLoadEvent event = new EnergyTileLoadEvent(this);
			MinecraftForge.EVENT_BUS.post(event);
			addedToEnergyNet=true;
		}
	}
	
	@Optional.Method(modid = "ic2")
	private void RemoveFromIC2EnergyNet() {
		if (!getWorld().isRemote) {
			MinecraftForge.EVENT_BUS.post(new EnergyTileUnloadEvent(this));
			addedToEnergyNet = false;
		}
	}

	@Override
	public Container getContainer(InventoryPlayer inventoryplayer) {
		return new ContainerForceEnergyExtractor(inventoryplayer.player, this);
	}

	@Override
	public void readExtraNBT(NBTTagCompound nbttagcompound) {
		super.readExtraNBT(nbttagcompound);
		ForceEnergybuffer = nbttagcompound.getInteger("ForceEnergybuffer");
		WorkEnergy = nbttagcompound.getInteger("WorkEnergy");
		WorkCylce = nbttagcompound.getInteger("WorkCylce");
	}

	@Override
	public void writeExtraNBT(NBTTagCompound nbttagcompound) {
		super.writeExtraNBT(nbttagcompound);
		nbttagcompound.setInteger("WorkCylce", WorkCylce);
		nbttagcompound.setInteger("WorkEnergy", WorkEnergy);
		nbttagcompound.setInteger("ForceEnergybuffer", ForceEnergybuffer);
	}

	@Override
	public String getName() {
		return "Extractor";
	}

	@Override
	public boolean hasCustomName() {
		return false;
	}

	@Override
	public List<String> getFieldsforUpdate() {
		List<String> NetworkedFields = new LinkedList<String>();
		NetworkedFields.clear();

		NetworkedFields.addAll(super.getFieldsforUpdate());
		NetworkedFields.add("capacity");
		NetworkedFields.add("WorkCylce");
		NetworkedFields.add("WorkEnergy");
		NetworkedFields.add("workdone");

		return NetworkedFields;
	}

	@Override
	public boolean isItemValid(ItemStack par1ItemStack, int Slot) {
		switch (Slot) {
		case 0:
			if ((par1ItemStack.getItem() instanceof ItemForcicium || par1ItemStack
					.getItem() instanceof ItemForcicumCell)
					&& getStackInSlot(4).isEmpty())
				return true;
			break;

		case 1:
			if (par1ItemStack.getItem() instanceof IPowerLinkItem)
				return true;
			break;

		case 2:
			if (par1ItemStack.getItem() instanceof ItemCapacitorUpgradeCapacity)
				return true;
			break;

		case 3:
			if (par1ItemStack.getItem() instanceof ItemExtractorUpgradeBooster)
				return true;
			break;

		case 4:
			if (par1ItemStack.getItem() instanceof ItemForcicumCell
					&& getStackInSlot(0).isEmpty())
				return true;
			break;
		}
		return false;
	}

	@Override
	public int getSlotStackLimit(int Slot) {
		switch (Slot) {
		case 0: // Forcicium
			return 64;
		case 1: // Powerlink
			return 1;
		case 2: // Cap upgrade
			return 9;
		case 3: // Boost upgrade
			return 19;
		case 4: // Forcicium cell
			return 1;
		}
		return 1;
	}




	@Override
	public void invalidate() {
		if (addedToEnergyNet && ModularForceFieldSystem.ic2Found) {
			RemoveFromIC2EnergyNet();
		}
		Linkgrid.getWorldMap(getWorld()).getExtractor().remove(getDeviceID());

		super.invalidate();
	}
	
	@Override
	public void onChunkUnload() {
		if (addedToEnergyNet && ModularForceFieldSystem.ic2Found) {
			RemoveFromIC2EnergyNet();
		}
		
		super.onChunkUnload();
	}

	public void converMJtoWorkEnergy() {
		/*if (this.getWorkEnergy() < this.getMaxWorkEnergy()) {
			float use = powerProvider
					.useEnergy(1, (float) (this.getMaxWorkEnergy() - this
							.getWorkEnergy() / 2.5), true);

			if (getWorkEnergy() + (use * 2.5) > getMaxWorkEnergy()) {
				setWorkEnergy(getMaxWorkEnergy());
			} else {
				setWorkEnergy((int) (getWorkEnergy() + (use * 2.5)));
			}
		}*/
	}

	/*@Override
	public void setPowerProvider(IPowerProvider provider) {
		this.powerProvider = provider;
	}

	@Override
	public IPowerProvider getPowerProvider() {
		return powerProvider;
	}

	@Override
	public void doWork() {
	}

	@Override
	public int powerRequest(ForgeDirection from) {
		double workEnergyinMJ = getWorkEnergy() / 2.5;
		double MaxWorkEnergyinMj = getMaxWorkEnergy() / 2.5;

		return (int) Math.round(MaxWorkEnergyinMj - workEnergyinMJ);
	}*/

	@Override
	public ItemStack getPowerLinkStack() {
		return this.getStackInSlot(getPowerlinkSlot());
	}

	@Override
	public int getPowerlinkSlot() {
		return 1;
	}

	@Override
	public TileEntityAdvSecurityStation getLinkedSecurityStation() {

		TileEntityCapacitor cap = Linkgrid.getWorldMap(getWorld()).getCapacitor()
				.get(getPowerSourceID());
		if (cap != null) {
			TileEntityAdvSecurityStation sec = cap.getLinkedSecurityStation();
			if (sec != null) {
				return sec;
			}

		}
		return null;
	}



	@Override
	@Optional.Method(modid = "redstoneflux")
	public boolean canConnectEnergy(EnumFacing from) {
		return true;
	}

	@Override
	@Optional.Method(modid = "redstoneflux")
	public int getEnergyStored(EnumFacing from) {
		return 0;
	}

	@Override
	@Optional.Method(modid = "redstoneflux")
	public int getMaxEnergyStored(EnumFacing from) {
		return 9999;
	}

	@Override
	@Optional.Method(modid = "redstoneflux")
	public int receiveEnergy(EnumFacing from, int maxReceive, boolean simulate) {

		double freeSpace = (double)(getMaxWorkEnergy() - getWorkEnergy());
		if(freeSpace==0)
		{

		}
		if(freeSpace >= maxReceive) {
			if(!simulate)
				setWorkEnergy(getWorkEnergy() + (int)maxReceive);
			return maxReceive;
		}
		else {
			if(!simulate)
				setWorkEnergy(getMaxWorkEnergy());
			return (int)(freeSpace);
		}
	}

	@Override
	public boolean hasCapability(Capability<?> capability, EnumFacing facing) {
		if(Loader.isModLoaded("tesla"))
			if (hasTeslaCapability(capability))
				return true;
		return super.hasCapability(capability, facing);
	}

	@Optional.Method(modid = "tesla")
	private boolean hasTeslaCapability(Capability<?> capability) {
		if(capability== TeslaCapabilities.CAPABILITY_CONSUMER)
			return true;
		return false;
	}

	@Override
	public <T> T getCapability(Capability<T> capability, EnumFacing facing) {
		if(Loader.isModLoaded("tesla"))
			if (hasTeslaCapability(capability))
				return TeslaCapabilities.CAPABILITY_CONSUMER.cast(cap);
		return super.getCapability(capability, facing);
	}


	@Override
	@Optional.Method(modid = "ic2")
	public double getDemandedEnergy() {

		if(!this.isActive())
			return 0;
		return (double)(getMaxWorkEnergy() - getWorkEnergy());
	}

	@Override
	@Optional.Method(modid = "ic2")
	public int getSinkTier() {
		return 3;
	}

	@Override
	@Optional.Method(modid = "ic2")
	public double injectEnergy(EnumFacing enumFacing, double v, double v1) {
		double freeSpace = (double) (getMaxWorkEnergy() - getWorkEnergy());
		if(getDemandedEnergy()<=0)
			return v;
		if (freeSpace >= v) {
			setWorkEnergy(getWorkEnergy() + (int) v);
			return 0;
		} else {
			setWorkEnergy(getMaxWorkEnergy());
			return (int) (v-freeSpace);
		}

	}

	@Override
	@Optional.Method(modid = "ic2")
	public boolean acceptsEnergyFrom(IEnergyEmitter iEnergyEmitter, EnumFacing enumFacing) {
		return true;
	}
}
